django/db/models/aggregates.py

```diff
@@ -36,6 +44,34 @@ def default_alias(self):
     def get_group_by_cols(self):
         return []
 
+    def get_case_expression(self):
+        field = self.get_source_expressions()[0]
+        return Case(
+            When(
+                self.filter,
+                then=field
+            )
+        )
+
+    def as_sql(self, compiler, connection, **extra_context):
+        if self.filter:
+            supports_filter = connection.features.supports_aggregate_filter_clause
+
+            if not supports_filter:
+                case_statement = self.get_case_expression()
+                self.set_source_expressions([case_statement])
```

Comment:
I think you should should avoid altering `self` in this case, you should clone `self` and call `set_source_expressions` on the clone.